apps/emqx_auth/src/emqx_authn/emqx_authn_chains.erl

```diff
@@ -747,7 +748,39 @@ maybe_add_stacktrace('throw', Data, _Stacktrace) ->
 maybe_add_stacktrace(_, Data, Stacktrace) ->
     Data#{stacktrace => Stacktrace}.
 
-authenticate_with_provider(#authenticator{id = ID, provider = Provider, state = State}, Credential) ->
+check_precondition(undefined, _) ->
+    true;
+check_precondition(Precondition, Credential0) ->
+    Credential = add_cert_info(Credential0),
+    case emqx_variform:render(Precondition, Credential) of
+        {ok, <<"true">>} ->
+            true;
+        Other ->
+            Other
+    end.
+
+%% Add cert info to credential if it exists to aid the authentication placeholders.
+add_cert_info(#{cn := CN, dn := DN} = Credential) ->
```

Comment:
Nit: 
Now we have some duplication with `emqx_auth_template:rename_client_info_vars/1`